### PR TITLE
Add actions computation

### PR DESCRIPTION
**Pull request by MarcBoelle3**
Separated refactor and add of docs in 2 different commits.

### CODE REVIEW

File: jumanji/environments/packing/job_shop/improvement/get_actions.py [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+    )  # shape (max_num_edges,)
+    senders_filtered = jnp.where(critical_blocks_mask, senders, -1)
+    receivers_filtered = jnp.where(critical_blocks_mask, receivers, -1)
+    sorted_edge_indices = jnp.argsort(filtered_est)
+    critical_block_pairs = jnp.stack(
+        [senders_filtered[sorted_edge_indices], receivers_filtered[sorted_edge_indices]], axis=-1
+    )
+
+    # === SECTION 3: Initialize Critical Block Information ===
+    critical_block_info = jnp.zeros((num_ops_total, 8), dtype=jnp.int32)
```
Comment (mintiti) [inline jumanji/environments/packing/job_shop/improvement/get_actions.py:119]: The docstring says it's (num_ops_total, 10), you have (num_ops_total, 8) here, which one is it ?

File: jumanji/environments/packing/job_shop/improvement/get_actions.py [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+    BLOCK_ID = 1  # the number of the critical block the operation belongs to
+    IS_LEFT = 2  # 1 if the operation is a left operation of a critical block, 0 otherwise
+    IS_RIGHT = 3  # 1 if the operation is a right operation of a critical block, 0 otherwise
+    LEFT_NEIGHBOR = 4  # the left neighbor of the operation
+    RIGHT_NEIGHBOR = 5  # the right neighbor of the operation
+    LEFT_END = 6  # operation index of the left end of the critical block
+    RIGHT_END = 7  # operation index of the right end of the critical block
+
+
+def get_critical_operations_features(
```
Comment (mintiti) [inline jumanji/environments/packing/job_shop/improvement/get_actions.py]: Those functions are pretty long, can we rewrite them slightly so the code is more self documenting?
From what I can see the function get_critical_operations_features could be refactored into:
* critical operation identification (isclose(est, lst))
* edge finding and sorting (finding machine constraint edges between critical ops)
* critical block initialization (setting up the initial block structure)
* critical block merging (the while loop that connects adjacent operations)
* block finalization (setting right ends and identifying left operations)
* time gap computation (calculating gaps between operations on same machine)

Same for the n6 function ?

File: jumanji/environments/packing/job_shop/improvement/get_actions.py [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+
+    # Obtain earliest start time of job successor of left end operation
+    is_left_end_last_of_job = (
+        left_end_idx % max_num_ops == num_ops_per_job[left_end_idx // max_num_ops] - 1
+    )
+    job_successor_of_left_end_idx = jnp.where(is_left_end_last_of_job, -1, left_end_idx + 1)
+    est_job_successor_of_left_end_idx = est_ops[job_successor_of_left_end_idx]
+
+    # Obtain earliest start time of job predecessor of current operation
+    is_current_op_first_of_job = ops_idx % max_num_ops == 0
```
Comment (mintiti) [inline jumanji/environments/packing/job_shop/improvement/get_actions.py]: (minor) can you use ops_order_in_job here in place of ops_idx % max_num_ops ?

```python
is_current_op_first_of_job = ops_order_in_job == 0
```

File: jumanji/environments/packing/job_shop/improvement/get_actions.py [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+    job_predecessor_of_right_end_idx = right_end_idx - 1
+    est_job_predecessor_of_right_end_idx = est_ops[job_predecessor_of_right_end_idx]
+
+    # Obtain earliest start time of job successor of current operation
+    is_current_op_last_of_job = ops_order_in_job == num_ops_per_job[ops_job_ids] - 1
+    job_successor_of_current_op_idx = jnp.where(is_current_op_last_of_job, -1, ops_idx + 1)
+    est_job_successor_of_current_op = est_ops[job_successor_of_current_op_idx]
+
+    # Sufficient condition for no cycle creation involving job predecessor of right end
+    job_predecessor_condition_satisfied = jnp.where(
```
Comment (mintiti) [inline jumanji/environments/packing/job_shop/improvement/get_actions.py:553]: I know this is complicated logic, any way we can give an idea/intuition on why this works ? 
code is a little bit obscure to read 😅 
Maybe something like this ? 
```python
# If current op's successor starts after right_end's predecessor,
# moving current op to right_end position won't create a backward edge
```

**Review by mintiti — COMMENTED** [review]
part 1

File: jumanji/environments/packing/job_shop/improvement/get_actions_test.py
Comment (mintiti) [inline jumanji/environments/packing/job_shop/improvement/get_actions_test.py:1]: I think there's a few tests that we could add here ?

1. The fundamental property that critical operations have est==lst isn't tested:
   - Test that all operations marked as critical actually satisfy est[op] == lst[op]
   - Test that sum of critical path durations equals makespan

2. We test that blocks exist but not their structural properties:
   - Test that blocks form valid linked lists via LEFT/RIGHT_NEIGHBOR
   - Test that blocks are maximal (no adjacent critical ops from different blocks on same machine)
   - Test that each critical op belongs to exactly one block

3. Currently we only test the action mask shape, not the safety properties:
   - Test that the EST comparisons (job_successor > job_predecessor_right_end, etc.) actually prevent cycles
   - Maybe apply each allowed N6 move and verify the graph remains acyclic?

4. Gap computation validation:
   - Test that gap_left + duration + gap_right spans the machine timeline correctly
   - Test edge cases like first/last operations on machines

6. Test the full pipeline from schedule to valid moves:
   - Verify that applying any valid N5/N6 move produces a valid schedule

**Review by mintiti — CHANGES_REQUESTED** [review]
Ok a few more comments for the tests. 
Other than that it looks good ! 
Good job!

File: jumanji/environments/packing/job_shop/improvement/get_actions_test.py
Comment (MarcBoelle3) [reply jumanji/environments/packing/job_shop/improvement/get_actions_test.py:1]: Done for 1. to 4. ! For 5., this test may be better in the env_test.py file, since it will be easier to step the environment and check that we obtain a valid schedule?

File: jumanji/environments/packing/job_shop/improvement/get_actions.py (hunk 4 above)
Comment (MarcBoelle3) [reply jumanji/environments/packing/job_shop/improvement/get_actions.py:553]: done!

File: jumanji/environments/packing/job_shop/improvement/get_actions.py (hunk 3 above)
Comment (MarcBoelle3) [reply jumanji/environments/packing/job_shop/improvement/get_actions.py]: done!

File: jumanji/environments/packing/job_shop/improvement/get_actions.py (hunk 2 above)
Comment (MarcBoelle3) [reply jumanji/environments/packing/job_shop/improvement/get_actions.py]: done!

File: jumanji/environments/packing/job_shop/improvement/get_actions.py (hunk 1 above)
Comment (MarcBoelle3) [reply jumanji/environments/packing/job_shop/improvement/get_actions.py:119]: typo here, done!

File: jumanji/environments/packing/job_shop/improvement/get_actions.py [hunk 5]
```diff
@@ -430,9 +682,79 @@ def get_action_mask_n6(
     )
 
     # Stack left and right validity into final action mask
-    action_mask = jnp.stack([left_move_valid, right_move_valid], axis=-1)
+    return jnp.stack([left_move_valid, right_move_valid], axis=-1).astype(jnp.bool_)
```
Comment (mintiti) [inline jumanji/environments/packing/job_shop/improvement/get_actions.py:685]: There's a slight difference here:
`.astype(jnp.bool_)` part is gone, not sure if it changes something

File: jumanji/environments/packing/job_shop/improvement/get_actions_test.py [hunk 6]
```diff
@@ -0,0 +1,1002 @@
...
+
+        # Test right move (direction 1)
+        action = jnp.array([1, 1], dtype=jnp.int32)  # op 1, direction right
+        result = select_operations_to_switch(critical_block_info, action, neighborhood=5)
+
+        assert result.shape == (3,)
+        assert result[0] == 1  # chosen operation comes first
+        assert result[2] == 1  # direction
+
+    def test_n6_operation_selection(
```
Comment (mintiti) [inline jumanji/environments/packing/job_shop/improvement/get_actions_test.py]: Both tests for n5 and n6 are very similar, is there a way to make it one parametrized test ? 
```python
# Could still benefit from parametrization:
@pytest.mark.parametrize("neighborhood,action,expected", [
    (5, [5, 0], [neighbor_of_5, 5, 0]),
    (6, [6, 0], [left_end_of_6, 6, 0]),
])
def test_operation_selection(self, neighborhood, action, expected, ...):
    # Single test for multiple cases
```

**Review by mintiti — APPROVED** [review]
LGTM, just a few minor things now, you can merge once you fixed them

File: jumanji/environments/packing/job_shop/improvement/get_actions.py (hunk 5 above)
Comment (MarcBoelle3) [reply jumanji/environments/packing/job_shop/improvement/get_actions.py:685]: It just enforces the types to be bool after the stack, but was already bool before
